api/src/main/java/io/grpc/Metadata.java

```diff
@@ -706,14 +813,36 @@ public String toString() {
      * @return a parsed instance of type T
      */
     abstract T parseBytes(byte[] serialized);
+
+    /**
+     * @return whether this key will be serialized to bytes lazily.
+     */
+    boolean serializesToStreams() {
+      return false;
+    }
+
+    /**
+     * Gets this keys (implementation-specific) marshaller, or null if the
+     * marshaller is not of the given type.
+     *
+     * @param marshallerClass The type we expect the marshaller to be.
+     * @return the marshaller object for this key, or null.
+     */
+    @Nullable
+    final <M> M getMarshaller(Class<M> marshallerClass) {
```

Comment:
Seems this method is only for binary stream marshaller, why not just
```java
    @Nullable
    @SuppressWarnings("unchecked")
    final BinaryStreamMarshaller<T> getBinaryStreamMarshaller() {
      if (BinaryStreamMarshaller.class.isInstance(marshaller)) {
        return (BinaryStreamMarshaller<T>) marshaller;
      }
      return null;
    }
```

Reply:
I wanted to avoid having the Key class directly depend on a single Marshaller type really. This approach let me avoid that, while supporting other marshaller types as well. If figured this approach was worth it since we'd discussed possibly adding a TextStreamMarshaller later on.

I'm happy either way though, so I'll defer to your preference.

Reply:
I saw the method and thought it was a bit odd, but it didn't see an obvious simple alternative that everyone would agree is superior. It is a bit weirder to be BinaryStreamMarshaller-specific as long as it is on the Key class. I'd be prone to moving the method to LazyStreamBinaryKey. But it just seemed beside the point and what is here is fine. We can always change it later.